### PR TITLE
Fix error messages

### PR DESCRIPTION
**Pull request by Smit-create**
https://github.com/lcompilers/lpython/pull/1936#discussion_r1234416502
#1952

### CODE REVIEW

File: src/libasr/asr_utils.h [hunk 1]
```diff
@@ -1266,15 +1266,15 @@ static inline std::string type_to_str_python(const ASR::ttype_t *t,
...
         }
         case ASR::ttypeType::Enum: {
             ASR::Enum_t* d = ASR::down_cast<ASR::Enum_t>(t);
-            return "enum " + std::string(symbol_name(d->m_enum_type));
+            return std::string(symbol_name(d->m_enum_type));
         }
         case ASR::ttypeType::Union: {
             ASR::Union_t* d = ASR::down_cast<ASR::Union_t>(t);
-            return "union " + std::string(symbol_name(d->m_union_type));
+            return std::string(symbol_name(d->m_union_type));
```
Comment (Shaikh-Ubaid) [inline src/libasr/asr_utils.h:1277]: I think this is not needed.

This was a feature added in https://github.com/lcompilers/lpython/commit/7f7dbde2f14a2ef244bd585f4dd69bc599efe8f5. `Struct`, `Enum`, `Union` are derived types. For them, printing just the symbol name does not seem sufficient (we also need their type information). 

For example, for `s: i32`, `type_to_str_python()` returns `i32` which makes clear that it is an integer/`i32`.
For `class S:`, `type_to_str_python()` returns just `S` which does not inform of the type of `S` (`Struct` or `Enum` or `Union`).

For other types, `type_to_str_python()` returns the type name, where as for derived types `type_to_str_python()` returns the symbol name (and not the type). Therefore, we had prepended the type information to derived types in https://github.com/lcompilers/lpython/commit/7f7dbde2f14a2ef244bd585f4dd69bc599efe8f5.

File: tests/reference/asr-structs_02-f95782c.stderr [hunk 2]
```diff
@@ -1,4 +1,4 @@
-semantic error: struct S s must be initialized a value
+semantic error: S s must be initialized a value
```
Comment (certik) [inline tests/reference/asr-structs_02-f95782c.stderr:1]:
```suggestion
semantic error: `s` must be initialized with an instance of S
```

File: tests/reference/asr-structs_02-f95782c.stderr [hunk 3]
```diff
@@ -1,4 +1,4 @@
...
  --> tests/errors/structs_02.py:8:5
   |
 8 |     s: S
```
Comment (certik) [inline tests/reference/asr-structs_02-f95782c.stderr:4]: And add a suggestion:
```
^
s: S = S()
```

File: tests/reference/asr-test_assign9-51278b8.stderr [hunk 4]
```diff
@@ -2,4 +2,4 @@ semantic error: Assignment to an input function parameter `x` is not allowed
  --> tests/errors/test_assign9.py:2:5
   |
 2 |     x = 2
-  |     ^ Use InOut[x] to allow assignment
+  |     ^ Assignment not supported for `x`
```
Comment (certik) [inline tests/reference/asr-test_assign9-51278b8.stderr:5]: For arrays, structs, etc the most common reason to get this error is when trying to modify the value in the caller. So the best suggestion is to use InOut (or Out).

For integers, people know that you cannot modify the variable in the caller that way, so I don't think that's the most common reason to get this error message.

I think there are two possibilities why somebody would assign to `x`:

* Trying to use the variable `x` locally in some algorithm and simply modifying it in the process, knowing that it doesn't change in the caller;
* Not needing the argument anymore and wanting to use a new local variable that happens to have the same name
* By a mistake wanting to use a new local variable that happens to shadow the argument, causing a bug

```suggestion
  |     ^ create a new local variable with a different name
```

**Review by certik — APPROVED** [review]

**Pull request merged**